File: stripe/src/test/java/com/stripe/android/model/StripeJsonUtilsTest.java
package com.stripe.android.model;

import com.stripe.android.testharness.JsonTestUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

/**
 * Test class for {@link StripeJsonUtils}.
 */
public class StripeJsonUtilsTest {

    private static final String SIMPLE_JSON_TEST_OBJECT =
            "{\n" +
                    "    \"akey\": \"avalue\",\n" +
                    "    \"bkey\": \"bvalue\",\n" +
                    "    \"boolkey\": true,\n" +
                    "    \"numkey\": 123\n" +
                    "}";

    private static final String NESTED_JSON_TEST_OBJECT =
            "{\n" +
                    "    \"top_key\": {\n" +
                    "        \"first_inner_key\": {\n" +
                    "            \"innermost_key\": 1000,\n" +
                    "            \"second_innermost_key\": \"second_inner_value\"\n" +
                    "        },\n" +
                    "        \"second_inner_key\": \"just a value\"\n" +
                    "    },\n" +
                    "    \"second_outer_key\": {\n" +
                    "        \"another_inner_key\": false\n" +
                    "    }\n" +
                    "}";

    private static final String SIMPLE_JSON_TEST_ARRAY = "[ 1, 2, 3, \"a\", true, \"cde\" ]";

    private static final String NESTED_MIXED_ARRAY_OBJECT =
            "{\n" +
                    "    \"outer_key\": {\n" +
                    "        \"items\": [\n" +
                    "            {\"id\": 123},\n" +
                    "            {\"id\": \"this time with letters\"},\n" +
                    "            \"a string item\",\n" +
                    "            256,\n" +
                    "            [ 1, 2, \"C\", 4],\n" +
                    "            [ {\"deep\": \"deepValue\"} ]\n" +
                    "        ],\n" +
                    "        \"another_key\": \"a simple value this time\"\n" +
                    "    },\n" +
                    "    \"other_outer_key\": false\n" +
                    "}";

    @Test
    public void nullIfNullOrEmpty_returnsNullForNull() {
        assertNull(StripeJsonUtils.nullIfNullOrEmpty("null"));
    }

    @Test
    public void nullIfNullOrEmpty_returnsNullForEmpty() {
        assertNull(StripeJsonUtils.nullIfNullOrEmpty(""));
    }

    @Test
    public void nullIfNullOrEmpty_returnsInputIfNotNull() {
        final String notANull = "notANull";
        assertEquals(notANull, StripeJsonUtils.nullIfNullOrEmpty(notANull));
    }

    @Test
    public void optString_whenFieldPresent_findsAndReturnsField() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("key", "value");
        assertEquals("value", StripeJsonUtils.optString(jsonObject, "key"));
    }

    @Test
    public void optString_whenFieldContainsRawNull_returnsNull() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("key", "null");
        assertNull(StripeJsonUtils.optString(jsonObject, "key"));
    }

    @Test
    public void optString_whenFieldNotPresent_returnsNull() throws JSONException {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("key", "value");
        Object ob = StripeJsonUtils.optString(jsonObject, "nokeyshere");
        assertNull(ob);
    }

    @Test
    public void jsonObjectToMap_forNull_returnsNull() {
        assertNull(StripeJsonUtils.jsonObjectToMap(null));
    }

    @Test
    public void jsonArrayToList_forNull_returnsNull() {
        assertNull(StripeJsonUtils.jsonArrayToList(null));
    }

    @Test
    public void jsonObjectToMap_forSimpleObjects_returnsExpectedMap() throws JSONException {
        Map<String, Object> expectedMap = new HashMap<>();
        expectedMap.put("akey", "avalue");
        expectedMap.put("bkey", "bvalue");
        expectedMap.put("boolkey", true);
        expectedMap.put("numkey", 123);

        JSONObject testJsonObject = new JSONObject(SIMPLE_JSON_TEST_OBJECT);
        Map<String, Object> mappedObject = StripeJsonUtils.jsonObjectToMap(testJsonObject);
        JsonTestUtils.assertMapEquals(expectedMap, mappedObject);
    }

    @Test
    public void jsonObjectToStringMap_forSimpleObjects_returnsExpectedMap() throws JSONException {
        Map<String, String> expectedMap = new HashMap<>();
        expectedMap.put("akey", "avalue");
        expectedMap.put("bkey", "bvalue");
        expectedMap.put("boolkey", "true");
        expectedMap.put("numkey", "123");

        JSONObject testJsonObject = new JSONObject(SIMPLE_JSON_TEST_OBJECT);
        Map<String, String> mappedObject =
                StripeJsonUtils.jsonObjectToStringMap(testJsonObject);
        JsonTestUtils.assertMapEquals(expectedMap, mappedObject);
    }

    @Test
    public void jsonObjectToMap_forNestedObjects_returnsExpectedMap() throws JSONException {
        Map<String, Object> expectedMap = new HashMap<>();
        expectedMap.put("top_key",
                new HashMap<String, Object>() {{
                    put("first_inner_key",
                            new HashMap<String, Object>() {{
                                put("innermost_key", 1000);
                                put("second_innermost_key", "second_inner_value");
                            }});
                    put("second_inner_key", "just a value");
                }});
        expectedMap.put("second_outer_key",
                new HashMap<String, Object>() {{
                    put("another_inner_key", false);
                }});

        JSONObject testJsonObject = new JSONObject(NESTED_JSON_TEST_OBJECT);
        Map<String, Object> mappedObject = StripeJsonUtils.jsonObjectToMap(testJsonObject);
        JsonTestUtils.assertMapEquals(expectedMap, mappedObject);
    }

    @Test
    public void jsonObjectToStringMap_forNestedObjects_returnsExpectedFlatMap()
            throws JSONException {
        Map<String, String> expectedMap = new HashMap<>();
        expectedMap.put("top_key", "{\"first_inner_key\":{\"innermost_key\":1000," +
                "\"second_innermost_key\":\"second_inner_value\"}," +
                "\"second_inner_key\":\"just a value\"}");
        expectedMap.put("second_outer_key", "{\"another_inner_key\":false}");

        JSONObject testJsonObject = new JSONObject(NESTED_JSON_TEST_OBJECT);
        Map<String, String> mappedObject =
                StripeJsonUtils.jsonObjectToStringMap(testJsonObject);
        JsonTestUtils.assertMapEquals(expectedMap, mappedObject);
    }

    @Test
    public void jsonObjectToMap_withNestedObjectAndArrays_returnsExpectedMap()
            throws JSONException {
        Map<String, Object> expectedMap = new HashMap<>();
        expectedMap.put("other_outer_key", false);

        final List<Object> itemsList = new ArrayList<>();
        itemsList.add(new HashMap<String, Object>() {{ put("id", 123); }});
        itemsList.add(new HashMap<String, Object>() {{
            put("id", "this time with letters");
        }});
        itemsList.add("a string item");
        itemsList.add(256);
        itemsList.add(Arrays.asList(1, 2, "C", 4));
        itemsList.add(Collections.singletonList(new HashMap<String, Object>() {{
            put("deep", "deepValue");
        }}));
        expectedMap.put("outer_key",
                new HashMap<String, Object>() {{
                    put("items", itemsList);
                    put("another_key", "a simple value this time");
                }});

        JSONObject testJsonObject = new JSONObject(NESTED_MIXED_ARRAY_OBJECT);
        Map<String, Object> convertedMap = StripeJsonUtils.jsonObjectToMap(testJsonObject);
        JsonTestUtils.assertMapEquals(expectedMap, convertedMap);
    }

    @Test
    public void jsonArrayToList_forSimpleList_returnsExpectedList() throws JSONException {
        List<Object> expectedList = new ArrayList<>();
        expectedList.add(1);
        expectedList.add(2);
        expectedList.add(3);
        expectedList.add("a");
        expectedList.add(true);
        expectedList.add("cde");

        JSONArray testJsonArray = new JSONArray(SIMPLE_JSON_TEST_ARRAY);
        List<Object> convertedJsonArray = StripeJsonUtils.jsonArrayToList(testJsonArray);
        JsonTestUtils.assertListEquals(expectedList, convertedJsonArray);
    }
}

File: stripe-core/src/test/java/com/stripe/android/core/model/StripeJsonUtilsTest.kt
package com.stripe.android.core.model

import org.json.JSONArray
import org.json.JSONObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * Test class for [StripeJsonUtils].
 */
class StripeJsonUtilsTest {

    @Test
    fun nullIfNullOrEmpty_returnsNullForNullString() {
        assertNull(StripeJsonUtils.nullIfNullOrEmpty("null"))
    }

    @Test
    fun nullIfNullOrEmpty_returnsNullForEmpty() {
        assertNull(StripeJsonUtils.nullIfNullOrEmpty(""))
    }

    @Test
    fun nullIfNullOrEmpty_returnsNullForNull() {
        assertNull(StripeJsonUtils.nullIfNullOrEmpty(null))
    }

    @Test
    fun nullIfNullOrEmpty_returnsValueForNonEmpty() {
        assertEquals("value", StripeJsonUtils.nullIfNullOrEmpty("value"))
    }

    @Test
    fun nullIfNullOrEmpty_returnsInputIfNotNull() {
        val notANull = "notANull"
        assertEquals(notANull, StripeJsonUtils.nullIfNullOrEmpty(notANull))
    }

    @Test
    fun optString_whenFieldPresent_findsAndReturnsField() {
        val jsonObject = JSONObject()
            .put("key", "value")
        assertEquals("value", StripeJsonUtils.optString(jsonObject, "key"))
    }

    @Test
    fun optString_whenFieldContainsRawNull_returnsNull() {
        val jsonObject = JSONObject()
            .put("key", "null")
        assertNull(StripeJsonUtils.optString(jsonObject, "key"))
    }

    @Test
    fun optString_whenFieldNotPresent_returnsNull() {
        val jsonObject = JSONObject()
            .put("key", "value")
        val ob = StripeJsonUtils.optString(jsonObject, "nokeyshere")
        assertNull(ob)
    }

    @Test
    fun jsonObjectToMap_forNull_returnsNull() {
        assertNull(StripeJsonUtils.jsonObjectToMap(null))
    }

    @Test
    fun jsonArrayToList_forNull_returnsNull() {
        assertNull(StripeJsonUtils.jsonArrayToList(null))
    }

    @Test
    fun jsonObjectToMap_forSimpleObjects_returnsExpectedMap() {
        val expectedMap = mapOf(
            "akey" to "avalue",
            "bkey" to "bvalue",
            "boolkey" to true,
            "numkey" to 123
        )

        val mappedObject = StripeJsonUtils.jsonObjectToMap(SIMPLE_JSON_TEST_OBJECT)
        assertEquals(expectedMap, mappedObject)
    }

    @Test
    fun jsonObjectToStringMap_forSimpleObjects_returnsExpectedMap() {
        val expectedMap = mapOf(
            "akey" to "avalue",
            "bkey" to "bvalue",
            "boolkey" to "true",
            "numkey" to "123"
        )

        val mappedObject = StripeJsonUtils.jsonObjectToStringMap(SIMPLE_JSON_TEST_OBJECT)
        assertEquals(expectedMap, mappedObject)
    }

    @Test
    fun jsonObjectToMap_forNestedObjects_returnsExpectedMap() {
        val expectedMap = mapOf(
            "top_key" to mapOf(
                "first_inner_key" to mapOf(
                    "innermost_key" to 1000,
                    "second_innermost_key" to "second_inner_value"
                ),
                "second_inner_key" to "just a value"
            ),
            "second_outer_key" to mapOf("another_inner_key" to false)
        )

        val mappedObject = StripeJsonUtils.jsonObjectToMap(NESTED_JSON_TEST_OBJECT)
        assertEquals(expectedMap, mappedObject)
    }

    @Test
    fun jsonObjectToStringMap_forNestedObjects_returnsExpectedFlatMap() {
        val expectedMap = mapOf(
            "top_key" to
                "{\"first_inner_key\":{\"innermost_key\":1000," +
                "\"second_innermost_key\":\"second_inner_value\"}," +
                "\"second_inner_key\":\"just a value\"}",
            "second_outer_key" to "{\"another_inner_key\":false}"
        )

        val mappedObject = StripeJsonUtils.jsonObjectToStringMap(NESTED_JSON_TEST_OBJECT)
        assertEquals(expectedMap, mappedObject)
    }

    @Test
    fun jsonObjectToMap_withNestedObjectAndArrays_returnsExpectedMap() {
        val items = listOf(
            mapOf("id" to 123),
            mapOf("id" to "this time with letters"),
            "a string item",
            256,
            listOf(1, 2, "C", 4),
            listOf(mapOf("deep" to "deepValue"))
        )
        val expectedMap = mapOf(
            "other_outer_key" to false,
            "outer_key" to mapOf(
                "items" to items,
                "another_key" to "a simple value this time"
            )
        )

        val convertedMap =
            StripeJsonUtils.jsonObjectToMap(NESTED_MIXED_ARRAY_OBJECT)
        assertEquals(expectedMap, convertedMap)
    }

    @Test
    fun jsonArrayToList_forSimpleList_returnsExpectedList() {
        val expectedList = listOf(1, 2, 3, "a", true, "cde")
        val convertedJsonArray = StripeJsonUtils.jsonArrayToList(SIMPLE_JSON_TEST_ARRAY)
        assertEquals(expectedList, convertedJsonArray)
    }

    private companion object {

        private val SIMPLE_JSON_TEST_OBJECT = JSONObject(
            """
            {
                "akey": "avalue",
                "bkey": "bvalue",
                "boolkey": true,
                "numkey": 123
            }
            """.trimIndent()
        )

        private val NESTED_JSON_TEST_OBJECT = JSONObject(
            """
            {
                "top_key": {
                    "first_inner_key": {
                        "innermost_key": 1000,
                        "second_innermost_key": "second_inner_value"
                    },
                    "second_inner_key": "just a value"
                },
                "second_outer_key": {
                    "another_inner_key": false
                }
            }
            """.trimIndent()
        )

        private val SIMPLE_JSON_TEST_ARRAY = JSONArray(
            """
            [1, 2, 3, "a", true, "cde"]
            """.trimIndent()
        )

        private val NESTED_MIXED_ARRAY_OBJECT = JSONObject(
            """
            {
                "outer_key": {
                    "items": [{
                            "id": 123
                        },
                        {
                            "id": "this time with letters"
                        },
                        "a string item",
                        256,
                        [1, 2, "C", 4],
                        [{
                            "deep": "deepValue"
                        }]
                    ],
                    "another_key": "a simple value this time"
                },
                "other_outer_key": false
            }
            """.trimIndent()
        )
    }
}
